### PR TITLE
[Core] `single_node_oom.aws`: Fix container mem stats and add randomness

### LINKED ISSUE #44393: [Core] CI: release test `single_node_oom.aws` flaky
### What happened + What you expected to happen

`single_node_oom.aws` release test failed since 03/10/2024:

image

Full dataset:
[20240401_184256.csv](https://github.com/ray-project/ray/files/14827347/20240401_184256.csv)

### Versions / Dependencies

Latest master

### Reproduction script

`single_node_oom.aws`

### Issue Severity

None

### PR DESCRIPTION
**Pull request by hongchaodeng**
## Why are these changes needed?

~~The test OOM-ed because the machine has 4 cpus and each test spins up 4 workers. Each worker is using 45% of memory. We cannot fit more than 2 workers on each node.~~

Another issue is that it is using `pstuil` to get memory stats. But release test is run in Anyscale job which is a containerized environment. We should use cgroup v2 files to get memory stats instead.

So this PR fix ~~the cpu/mem calculation and~~ the memory stats query method.

I have tested in release environments and it works many times:

image

## Related issue number

Closes https://github.com/ray-project/ray/issues/44393

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: release/nightly_tests/stress_tests/test_parallel_tasks_memory_pressure.py [hunk 1]
```diff
@@ -19,10 +19,23 @@ def allocate_memory(
...
+def get_container_memory_limit():
+    with open("/sys/fs/cgroup/memory.max") as f:
+        limit_bytes = int(f.read().strip())
+    return limit_bytes
+
+
+def get_container_memory_usage():
+    with open("/sys/fs/cgroup/memory.current") as f:
+        usage_bytes = int(f.read().strip())
+    return usage_bytes
```
Comment (jjyao) [inline release/nightly_tests/stress_tests/test_parallel_tasks_memory_pressure.py:31]: We have existing `get_system_memory()` and `get_used_memory()` that works for cgroup. You can just call those.

Comment (hongchaodeng) [reply release/nightly_tests/stress_tests/test_parallel_tasks_memory_pressure.py:31]: Nice! Let me refactor it

**Review by hongchaodeng — COMMENTED** [review]
Sure! Will update soon

**Comment by hongchaodeng** [comment]
@jjyao Fixed.
I also added a logic in the test code that will wait a bit more time as it is approaching the OOM threshold. This will make the test more stable despite its flaky nature.

**Comment by hongchaodeng** [comment]
After digging into the logs, I have a second thought about what's happened and caused the issue.

First of all, the test is flaky by nature. There are 4 tasks, each consuming nearly 45% memory in 10 steps (4.5% per step). After 6 steps, each tasks will consume ~22.5%, sum to ~90%. For the next step, if 3 tasks try to allocate memory, that would be ~97.5. At this point, we rely on the memory monitor to kick in and kill tasks. (This has been observed in the logs.)

But the issue is that there might still be race conditions. If memory monitor didn't kill task fast enough, or tasks were allocating memories too fast, then the node OOM will kick in and kill Ray components and cause test failures. (This has also been observed in the logs.)

Instead of sleeping for 1s for all tasks, I added some randomness to avoid all workers allocating memories at the same time after mid point. This will make some room for memory monitor to kill tasks and prevent all tasks allocating memory at the same time.

**Review by jjyao — APPROVED** [review]

**Pull request merged**